src/v/cluster/seq_stm.cc

```diff
@@ -236,61 +105,29 @@ seq_stm::replicate(
             _seq_table.emplace(entry.pid, entry);
         } else {
             if (!is_sequence(pid_seq->second.seq, bid.first_seq)) {
-                return seastar::make_ready_future<
-                  checked<raft::replicate_result, kafka::error_code>>(
+                co_return checked<raft::replicate_result, kafka::error_code>(
                   kafka::error_code::out_of_order_sequence_number);
             }
-            pid_seq->second.seq += bid.last_seq;
+            pid_seq->second.seq = bid.last_seq;
```

Comment:
In here we update internal state of `seq_stm` before replicate finished. I assume that this is safe to be done since we return error to client ?

Reply:
I designed it this way and you're right it isn't 100% right thing to do but the worst thing that may happen is that in a rare situation we reject a write that otherwise would have passed. But client expects this behaviour and it terminates session and starts a new session so nothing is violated.